compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs

```diff
@@ -162,14 +162,16 @@ pub trait HirTyLowerer<'tcx> {
     /// Record the lowered type of a HIR node in this context.
     fn record_ty(&self, hir_id: HirId, ty: Ty<'tcx>, span: Span);
 
-    /// The inference context of the lowering context if applicable.
-    fn infcx(&self) -> Option<&InferCtxt<'tcx>>;
+    /// The inference context of the lowering context.
+    fn infcx(&self) -> &InferCtxt<'tcx>;
```

Comment:
Yeah, I'd prefer if we don't have an `InferCtxt` in `ItemCtxt`, since any inference performed in the `ItemCtxt` is going to be lost to the void, and almost all cases where `ItemCtxt` passes back a hir-lowered value, it's not correct to reference infer regions, etc.

The `InferCtxt` used by `ItemCtxt` is only for very specific cases, such a diagnostics, and for places that should eventually probably be reworked (or at least revised with a lot of scrutiny) like inherent-associated-types.